lib/fluent/plugin/in_tail.rb

```diff
@@ -120,7 +126,25 @@ def initialize
       config_argument :usage, :string, default: 'in_tail_parser'
     end
 
-    attr_reader :paths
+    config_section :group, param_name: :group, required: false, multi: false do
+      desc 'Regex for extracting group\'s metadata'
+      config_param :pattern, 
+                   :regexp, 
+                   default: /var\/log\/containers\/(?<appname>[a-z0-9]([-a-z0-9]*[a-z0-9])?(\/[a-z0-9]([-a-z0-9]*[a-z0-9])?)*)_(?<namespace>[^_]+)_(?<container>.+)-(?<docker_id>[a-z0-9]{64})\.log$/
+      desc 'Period of time in which the group_line_limit is applied'
+      config_param :rate_period, :time, default: 5
+
+      config_section :rule, multi: true, required: true do
+        desc 'Namespace key'
+        config_param :namespace, :array, value_type: :string, default: [DEFAULT_NAMESPACE]
+        desc 'App name key'
+        config_param :appname, :array, value_type: :string, default: [DEFAULT_APPNAME]
```

Comment:
I'm considering how to generalize this part.
As I mentioned before, I think this part has following 2 issues:

* Depends on k8s's concept
* config_param `appname` and `namespace` are hardcoded while `pattern` can be modified by user

The former one is acceptable for me since it's only a default value, it's customizable (and this default value is appropriate since k8s is most major application recent days).
But the later one is a bit uncomfortable for me.

One idea I'm considering is extracting named caputures from `pattern` (by `pattern.named_captures`) and generate `config_param` dynamically. But it might too complicate.

Do you have any other idea to resolve it? @Pranjal-Gupta2 @cosmo0920

Reply:
Or using hash & array param might be enough for this case

Reply:
Thanks for your suggestions @ashie! 
I was also thinking in the direction of generating `config_param` dynamically. 

```ruby
      config_section :rule, multi: true, required: true do 
        Regexp.compile(conf.elements('group')[0]['pattern']).named_captures.keys.each do |key|
          config_param key.to_sym, :array, value_type: :string, default: [DEFAULT_APPNAME]
        end
        desc 'Maximum number of log lines allowed per group over a period of rate_period'
        config_param :limit, :integer, default: DEFAULT_LIMIT
      end
```

I added this snippet in the `configure` method just before calling `super` ([in_tail.rb#L163](https://github.com/fluent/fluentd/blob/514126ca4329c81739925aa6bd140fabe87faaf2/lib/fluent/plugin/in_tail.rb#L163)). However, I am getting this error: 
```bash
/.rvm/gems/ruby-3.0.0/gems/fluentd-1.14.4/lib/fluent/plugin/in_tail.rb:160:in `configure': undefined method `config_section' for #<Fluent::Plugin::TailInput:000000000010b8> (NoMethodError)
Did you mean?  configured_section_create
```

If we can call `config_section` and `config_param` in the `configure` method then I believe we could generate dynamic parameters for our plugin. What do you think? @ashie @cosmo0920

Reply:
Now I think generating `config_param` dynamically is too complicate.
How about using `:hash` type like this?:

```ruby
config_param :match, :hash, default: {namespace: [DEFAULT_NAMESPACE], appname: [DEFAULT_APPNAME]}
```

usage:
```xml
<rule>
  match {
    "namespace": ["aaa", "bbb"],
    "appname": ["ccc", "ddd"]
  }
  limit 20
</rule>
```

It would be better that single string is also acceptable:
```xml
<rule>
  match {
    "namespace": "namespace1",
    "appname": "app1"
  }
 limit 20
</rule>
```

Reply:
Now I think that generating `config_params` is too complicate and probably too hard or impossible for this case.
I'm considering another idea, using `:hash` type of `config_param` like this:

```ruby
    config_param :match, :hash, default: {namespace: [DEFAULT_NAMESPACE], appname: [DEFAULT_APPNAME}, value_type: :xxx
```

Although current implementation can't accept an array as a value of hash, I think there are 2 ways to resolve it.

1. Do not use array, express multiple items by regexp
2. Add a new `value_type` like `:string_array`

Example usage for 1.:
```xml
<rule>
  match {
         "namespace": "/spac1|space2|scape3/", 
         "appname": "/app1|app2|app3/"
       }
  ])
  limit 20
</rule>
```

Example patch & usage for 2.:

```diff
diff --git a/lib/fluent/config/types.rb b/lib/fluent/config/types.rb
index 98e7d0d9..b6836b8a 100644
--- a/lib/fluent/config/types.rb
+++ b/lib/fluent/config/types.rb
@@ -191,6 +191,10 @@ module Fluent
         when :time then Config.time_value(value, opts, name)
         when :regexp then Config.regexp_value(value, opts, name)
         when :symbol then Config.symbol_value(value, opts, name)
+        when :string_array
+          child_opts = opts.dup
+          child_opts[:value_type] = :string
+          Config.array_value(value, child_opts, name)
         else
           raise "unknown type in REFORMAT: #{type}"
         end
```

Example usage for 2.:
```xml
<rule>
  match {
         "namespace": ["spac1", "space2", "scape3"], 
         "appname": ["app1, "app2", "app3"]
       }
  ])
  limit 20
</rule>
```

Reply:
I think we can use a hash with `regexp` key values for defining rules. It will be easier to assign a group to a file using regular expressions.

Reply:
@ashie I have renamed `appname` to `podname`. So, now the default values of group are k8s specific but can be overridden in conf for other use cases.